File: src/frameworks/services/scrapingService/scrapers/HanjinScraper.ts
import * as cheerio from "cheerio"
import DeliveryLocationVO from "@domains/vos/DeliveryLocationVO"
import IDeliveryDTO from "@domains/dtos/interfaces/IDeliveryDTO"
import ILayerDTO from "@domains/dtos/interfaces/ILayerDTO"
import DeliveryProgressVO from "@domains/vos/DeliveryProgressVO"
import DeliveryDTO from "@domains/dtos/DeliveryDTO"
import LayerDTO from "@domains/dtos/LayerDTO"
import IServerHTTP from "@frameworks/infrastructures/interfaces/IServerHTTP"
import DeliveryStateGenerator from "../helpers/DeliveryStateGenerator"
import StringHelper from "../helpers/StringHelper"
import IScraper from "./interfaces/IScraper"

export default class HanjinScraper implements IScraper {
  constructor(private readonly serverHTTP: IServerHTTP) {}

  async getTrack(trackingNumber: string): Promise<ILayerDTO<IDeliveryDTO>> {
    try {
      const trackingRes = await this.serverHTTP.post(
        `https://www.hanjin.com/kor/CMS/DeliveryMgr/WaybillResult.do?wblnum=${trackingNumber}&mCode=MN038&schLang=KR`,
        {},
        {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded"
          }
        }
      )

      if (trackingRes.status !== 200) {
        return new LayerDTO({
          isError: true,
          message: "운송장 조회에 실패하였습니다."
        })
      }

      const resData = await trackingRes.text()
      const $ = cheerio.load(resData)
      const $wrap = $("#delivery-wr")
      if ($wrap.length === 0) {
        return new LayerDTO({
          isError: true,
          message: "해당 운송장이 존재하지 않거나 조회할 수 없습니다."
        })
      }

      const $informationTable = $wrap.find(".delivery-tbl").find("tbody")
      const $progressTable = $wrap.find(".waybill-tbl").find("table")
      const $informations = $informationTable.find("td")

      const progressVOs = []
      $progressTable
        .find("tbody")
        .find("tr")
        .each((_, element) => {
          const td = $(element).find("td")
          const description = StringHelper.trim(td.eq(3).text())
          const location = StringHelper.trim(td.eq(2).text())
          const time = this.parseDateTime(td.eq(0).text(), td.eq(1).text())
          const state = this.parseStatus(td.eq(3).text())
          progressVOs.push(
            new DeliveryProgressVO({
              description,
              location,
              time,
              state
            })
          )
        })
      progressVOs.reverse()

      const stateVO =
        progressVOs.length > 0
          ? progressVOs[0].state
          : this.parseStatus("상품준비중")

      const fromVO = new DeliveryLocationVO({
        name: this.parseLocationName($informations.eq(1).text()),
        time:
          progressVOs.length > 0 ? progressVOs[progressVOs.length - 1].time : ""
      })

      const toVO = new DeliveryLocationVO({
        name: this.parseLocationName($informations.eq(2).text()),
        time: stateVO.name === "배달완료" ? progressVOs[0].time : ""
      })

      const deliveryDTO = new DeliveryDTO({
        from: fromVO,
        to: toVO,
        progresses: progressVOs,
        state: stateVO
      })

      return new LayerDTO({
        data: deliveryDTO
      })
    } catch (error) {
      return new LayerDTO({
        isError: true,
        message: error.message
      })
    }
  }

  private parseLocationName(value: string) {
    const short = value.substring(0, 4)
    return short + (short.includes("*") ? "" : "*")
  }

  private parseDateTime(date: string, time: string) {
    return date + " " + time + ":00"
  }

  private parseStatus(value?: string) {
    if (typeof value !== "string") {
      return DeliveryStateGenerator.getState("상품이동중")
    }
    if (value.includes("집하")) {
      return DeliveryStateGenerator.getState("상품인수")
    }
    if (value.includes("배송출발")) {
      return DeliveryStateGenerator.getState("배달출발")
    }
    if (value.includes("배송완료")) {
      return DeliveryStateGenerator.getState("배달완료")
    }
    return DeliveryStateGenerator.getState("상품이동중")
  }
}

File: scraper/hanjin_scraper.go
package scraper

import (
	"errors"
	"fmt"
	"net/http"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/falsy/delivery-tracker-server/model"
)

func parseHanjinDateTime(date, time string) string {
	return strings.TrimSpace(date) + " " + strings.TrimSpace(time) + ":00"
}

func parseHanjinStatus(value string) model.DeliveryState {
	switch {
	case strings.Contains(value, "집하"):
		return model.DeliveryState{ID: "item_received", Name: "상품인수"}
	case strings.Contains(value, "배송출발"):
		return model.DeliveryState{ID: "out_for_delivery", Name: "배달출발"}
	case strings.Contains(value, "배송완료"):
		return model.DeliveryState{ID: "delivered", Name: "배달완료"}
	default:
		return model.DeliveryState{ID: "in_transit", Name: "상품이동중"}
	}
}

func HanjinGetTrack(trackingNumber string) (*model.DeliveryResult, error) {
	url := fmt.Sprintf("https://www.hanjin.com/kor/CMS/DeliveryMgr/WaybillResult.do?wblnum=%s&mCode=MN038&schLang=KR", trackingNumber)
	req, err := http.NewRequest("POST", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	if res.StatusCode != 200 {
		return nil, errors.New("운송장 조회에 실패하였습니다")
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	wrap := doc.Find("#delivery-wr")
	if wrap.Length() == 0 {
		return nil, errors.New("해당 운송장이 존재하지 않거나 조회할 수 없습니다")
	}

	infoTds := wrap.Find(".delivery-tbl").Find("tbody").Find("td")
	progressTable := wrap.Find(".waybill-tbl").Find("table")

	var progresses []model.DeliveryProgress
	progressTable.Find("tbody").Find("tr").Each(func(i int, s *goquery.Selection) {
		td := s.Find("td")
		description := strings.TrimSpace(td.Eq(3).Text())
		location := strings.TrimSpace(td.Eq(2).Text())
		time := parseHanjinDateTime(td.Eq(0).Text(), td.Eq(1).Text())
		state := parseHanjinStatus(description)

		progresses = append([]model.DeliveryProgress{{
			Description: description,
			Location:    location,
			Time:        time,
			State:       state,
		}}, progresses...)
	})

	state := parseHanjinStatus("상품이동중")
	if len(progresses) > 0 {
		state = progresses[0].State
	}

	from := model.DeliveryLocation{
		Name: parseLocationName(strings.TrimSpace(infoTds.Eq(1).Text())),
		Time: func() string {
			if len(progresses) > 0 {
				return progresses[len(progresses)-1].Time
			}
			return ""
		}(),
	}

	to := model.DeliveryLocation{
		Name: parseLocationName(strings.TrimSpace(infoTds.Eq(2).Text())),
		Time: func() string {
			if state.Name == "배달완료" && len(progresses) > 0 {
				return progresses[0].Time
			}
			return ""
		}(),
	}

	return &model.DeliveryResult{
		From:       from,
		To:         to,
		Progresses: progresses,
		State:      state,
	}, nil
}
